File: x.js
var delimiters = {};
delimiters["("] = true; delimiters[")"] = true;
delimiters[";"] = true; delimiters["\n"] = true;

var whitespace = {};
whitespace[" "] = true; whitespace["\t"] = true; whitespace["\n"] = true;

var operators = {};
operators["+"] = "+"; operators["<"] = "<";
operators["and"] = "&&"; operators["or"] = "||";

function makeStream(str) {
    return { pos: 0, string: str, len: str.length };
}

function peekChar(s) {
    if (s.pos < s.len)
        return s.string.charAt(s.pos);
}

function readChar(s) {
    var c = peekChar(s);

    if (c) {
        s.pos++;
        return c;
    }
}

function readWhitespace (s) {
    var c;

    while (true) {
        c = peekChar(s);
        if (c && whitespace[c])
            readChar(s);
        else break;
    }
}

function readAtom(s) {
    var c, str = "";

    while (true) {
        c = peekChar(s);
        if (c && !whitespace[c] && !delimiters[c]) {
            str += c;
            readChar(s);
        } else break;
    }

    return str;
}

function readList(s) {
    readChar(s); // (

    var c, l = [];

    while (true) {
        readWhitespace(s);
        c = peekChar(s)
        if (c && c != ")") {
            l.push(read(s));
        } else if (c) {
            readChar(s); // )
            break;
        } else {
            error("Expected ) at " + s.pos);
        }
    }

    return l;
}

function readComment(s) {
    var c;
    do { c = readChar(s); } while (c && c != "\n");
}

function readString(s) {
    readChar(s); // "

    var c, str = "";

    while (true) {
        c = peekChar(s);
        if (c && c != "\"") {
            if (c == "\\")
                str += readChar(s);
            str += readChar(s);
        } else if (c) {
            readChar(s); // "
            break;
        } else {
            error("Expected \" at " + s.pos);
        }
    }

    return str;
}

function read(s) {
    readWhitespace(s);

    var c = peekChar(s);

    switch (c) {
        case ";": readComment(s); return read(s);
        case "(": return readList(s);
        case ")": error("Unexpected ) at " + s.pos);
        case "\"": return readString(s);
        default: return readAtom(s);
    }
}

function readFromString(str) {
    return read(makeStream(str));
}

function isAtom(form) {
    return typeof form == "string";
}

function isCall(form) {
    return Array.isArray(form) && typeof form[0] == "string";
}

function isOperator(form) {
    return operators[form[0]] != null;
}

function compileAtom(form, isStatement) {
    return form.toString() + (isStatement ? ";" : "");
}

function compileCall(form, isStatement) {
    var i = 1, args = "";
    while (i < form.length) {
        var a = compile(form[i], false);
        args += a + (i < form.length - 1 ? "," : "");
        ++i;
    }
    return form[0] + "(" + args + ")" + (isStatement ? ";" : "");
}

function compileOperator(form) {
    if (form.length > 3)
        error("Too many arguments to operator " + form[0]);

    var a = compile(form[1], false);
    var b = compile(form[2], false);
    return "(" + a + operators[form[0]] + b + ")";
}

function compile(form, isStatement) {
    isStatement = typeof isStatement == "undefined" ? true : isStatement;

    if (isAtom(form))
        return compileAtom(form, isStatement);
    else if (isCall(form))
        if (isOperator(form)) {
            if (isStatement)
                error("Cannot compile use of operator " + form[0] + " as a statement");
            return compileOperator(form);
        } else return compileCall(form, isStatement);
    else error("Unexpected form " + form.toString());
}

function error(msg) {
    throw new Error(msg);
}

function test(actual, expected) {
    if (expected !== actual) {
        error("Expected " + expected + ", was " + actual);
    }
}

function runTests() {
    test(readFromString("()").length, [].length);
    test(readFromString("a"), "a");
    test(readFromString("(a b c)").length, 3);
    test(readFromString("(a b c) ; blaz").length, 3);
    test(readFromString("; blaz\n(a b c)").length, 3);
    test(readFromString("( 1  b delta)").length, 3);
    test(readFromString(""), "");
    test(readFromString("\"foo\""), "foo");
    test(readFromString("\"foo\n\""), "foo\n");
    test(readFromString("(\"a\" b c)")[0], "a");
}

File: x.lua
delimiters = {["("]=true, [")"]=true, [";"]=true, ["\n"]=true}
whitespace = {[" "]=true, ["\t"]=true, ["\n"]=true}

function make_stream (str)
  return { pos=1, string=str, len=string.len(str) }
end

function peek_char (s)
  if s.pos <= s.len then
    return string.sub(s.string, s.pos, s.pos)
  else
    return nil
  end
end

function read_char (s)
  local c = peek_char(s)
  if c then
    s.pos = s.pos+1
  end
  return c
end

function unread_char (s)
  if s.pos > 1 then
    s.pos = s.pos-1
  end
end

function read_whitespace (s)
  local c
  while true do
    c = peek_char(s)
    if c and whitespace[c] then
      read_char(s)
    else
      return
    end
  end
end

function read_atom (s)
  local c
  local str = ""
  while true do
    c = peek_char(s)
    if c and not whitespace[c] and not delimiters[c] then
      str = str .. c
      read_char(s)
    else
      return str
    end
  end
end

function read_list (s)
  read_char(s) -- (
  local l = {}
  local c
  while true do
    -- read_whitespace(s)
    c = peek_char(s)
    if c and c ~= ")" then
      table.insert(l, read(s))
    elseif not c then
      print("Expected ) at ", s.pos)
      error()
    else
      read_char(s) -- )
      break
    end
  end
  return l
end

function read_comment (s)
  local c
  repeat
    c = read_char (s)
  until c == "\n" or not c
end

function read_from_string (str)
  return read(make_stream(str))
end

function read (s)
  read_whitespace(s)
  local c = peek_char(s)
  if c == ";" then
    read_comment(s)
    return read(s)
  elseif c == "(" then
    return read_list(s)
  elseif c == ")" then
    print("Unexpected ) at", s.pos)
    error()
  else
    return read_atom(s)
  end
end
